Copy nibabel header when creating output prediction

jcohenadad opened the pull request:
## Checklist

#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [ ] I've assigned a reviewer

<!-- For the title, please observe the following rules:
	- Provide a concise and self-descriptive title
	- Do not include the applicable issue number in the title, do it in the PR body
	- If the PR is not ready for review, convert it to a draft.
-->

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
Before, the output prediction was created by instantiating a new nibabel object, via the copy of the original affine matrix. However, doing so does *not* ensure the same qform_code and sform_code are preserved. This PR fixes the issue.
More context: https://nipy.org/nibabel/nifti_images.html#default-sform-and-qform-codes

Testing for the proposed change will be implemented in a subsequent PR: https://github.com/ivadomed/ivadomed/issues/713

This PR also includes minor fixes to the unit testing documentation.

## Linked issues
Fixes #711
Related to https://github.com/neuropoly/spinalcordtoolbox/issues/3284, https://github.com/ivadomed/ivadomed/issues/720

jcohenadad committed:
Copy nibabel header when creating output prediction

Before, the output prediction was created by instantiating a new nibabel object, via the copy of the original affine matrix. However, doing so does *not* ensure the same qform_code and sform_code are preserved. This commit fixes the issue.

More context: https://nipy.org/nibabel/nifti_images.html#default-sform-and-qform-codes

Fixes #711

jcohenadad committed:
Fixed typo

jcohenadad committed:
Removed the additional "cd ivadomed"

Step 1 already includes cd to the ivadomed root repos. Another cd ivadomed will make the pytest fail (ie: nothing to test).

jcohenadad committed:
Fixed copying of header for nibabel objects with different size

Fixes https://github.com/ivadomed/ivadomed/pull/714/checks?check_run_id=2108027002#step:6:141

charleygros commented on ivadomed/inference.py:
```diff
@@ -86,7 +86,16 @@ def pred_to_nib(data_lst, z_lst, fname_ref, fname_out, slice_axis, debug=False,
         arr_pred_ref_space = postpro.apply()
-    nib_pred = nib.Nifti1Image(arr_pred_ref_space, nib_ref.affine)
+    if arr_pred_ref_space.shape == nib_ref.get_fdata().shape:
+        # Here we prefer to copy the header (rather than just the affine matrix), in order to preserve the qform_code.
+        # See: https://github.com/ivadomed/ivadomed/issues/711
```
Sounds great! and well documented, thanks

charleygros reviewed (commented):
Mmmm thank you for solving this... very specific issue!!

Less critical, but we may also want to apply this change to other instances in ivadomed?
- [uncertainty](https://github.com/ivadomed/ivadomed/blob/87d6a63d0bb0114bb3d3d8e13584089bf2ce0142/ivadomed/uncertainty.py#L76)
- [preprocessing](https://github.com/ivadomed/ivadomed/blob/3c9d32c41ef4222848f0dfb024e89c11f7264046/ivadomed/preprocessing.py#L36)
- [postprocessing](https://github.com/ivadomed/ivadomed/blob/3004dfe53afcc04529380ac419370ccb76f3a128/ivadomed/postprocessing.py#L26)
- [visualize](https://github.com/ivadomed/ivadomed/blob/a8f1839ace712347f046a4131cdd75a8c6cc43da/ivadomed/visualize.py#L226)
- [loader/adaptive](https://github.com/ivadomed/ivadomed/blob/fb3c913ddc3fb1942884bdbb16a7796e94d20192/ivadomed/loader/adaptative.py#L753)
- [evaluation](https://github.com/ivadomed/ivadomed/blob/87156e33be17eb96edb124e01e98d57619ea8a59/ivadomed/evaluation.py#L84)

jcohenadad commented:
> Mmmm thank you for solving this... very specific issue!!
> 
> Less critical, but we may also want to apply this change to other instances in ivadomed?
> 
> * [uncertainty](https://github.com/ivadomed/ivadomed/blob/87d6a63d0bb0114bb3d3d8e13584089bf2ce0142/ivadomed/uncertainty.py#L76)
> * [preprocessing](https://github.com/ivadomed/ivadomed/blob/3c9d32c41ef4222848f0dfb024e89c11f7264046/ivadomed/preprocessing.py#L36)
> * [postprocessing](https://github.com/ivadomed/ivadomed/blob/3004dfe53afcc04529380ac419370ccb76f3a128/ivadomed/postprocessing.py#L26)
> * [visualize](https://github.com/ivadomed/ivadomed/blob/a8f1839ace712347f046a4131cdd75a8c6cc43da/ivadomed/visualize.py#L226)
> * [loader/adaptive](https://github.com/ivadomed/ivadomed/blob/fb3c913ddc3fb1942884bdbb16a7796e94d20192/ivadomed/loader/adaptative.py#L753)
> * [evaluation](https://github.com/ivadomed/ivadomed/blob/87156e33be17eb96edb124e01e98d57619ea8a59/ivadomed/evaluation.py#L84)

Excellent point-- issue opened here https://github.com/ivadomed/ivadomed/issues/715

jcohenadad committed:
Only testing for nx,ny,nz in case the prediction shape is 4d

Added a TODO because this hack can break things.

jcohenadad committed:
Implemented the fix for #711 in split_classes()

This additional fix was necessary because segment_volume() calls this function at the end of the prediction.

jcohenadad committed:
Merge branch 'master' into jca/711-qform

andreanne-lemay commented on ivadomed/inference.py:
```diff
@@ -86,7 +86,23 @@ def pred_to_nib(data_lst, z_lst, fname_ref, fname_out, slice_axis, debug=False,
+    #  I'm afraid this [0:3] hack will break for 2D data. Also, we might want to make sure the output arr_pred_ref_space
+    #  always has the same shape as the input image. For multiple class prediction, maybe we can find another way
+    #  to store the predictions (e.g. list of np arrays?).
+    if arr_pred_ref_space.shape[0:3] == nib_ref.get_fdata().shape[0:3]:
```
From what I understand in the description above, if the dimensions of the prediction is (nx, ny, nz, 1) while the input image is (nx, ny, nz) the code should go to the "else" statement (correct me if I'm wrong). It seems to me that in all cases the code under `if arr_pred_ref_space.shape[0:3] == nib_ref.get_fdata().shape[0:3]:` will be ran since `[nx, ny, nz, 1][0:3] == [nx, ny, nz][[0:3] `. Maybe this bit of code is in the case of a prediction with the dimensions (1, nx, ny, nz)? 

From this error https://github.com/ivadomed/ivadomed/pull/714/checks?check_run_id=2107966525, the problem seems to come from the `save_color_labels` function which generates nifti files with non-int or float data. Rather than introducing this condition that might cause errors with 2D images, I'd simply comment the lines related to this function (which simply creates an extra file that we barely use) and adjust the function `save_color_labels`  in a future PR. If this plans work for you, I can take care of it.

jcohenadad replied:
> if the dimensions of the prediction is (nx, ny, nz, 1) while the input image is (nx, ny, nz) the code should go to the "else" statement (correct me if I'm wrong). 

sorry i wasn't clear-- let me know if this is clearer-- if it is i will update the code with this comment below:
~~~
    # TODO: remove the "[0:3]" in "if test" below. The reason I added it is because in some cases, the
    #  prediction has the shape (nx, ny, nz, 1) while the input image is (nx, ny, nz), even for a one-channel
    #  prediction. Consequently, the "if check" below goes to the "else" case, and
    #  issue https://github.com/ivadomed/ivadomed/issues/711 is still happening.
    #  What we want is: if the output prediction has the same (nx, ny, nz) as the reference image, the we need
    #  to copy the header (not just the affine), in order to fix #711.
    #  However, I am afraid this "[0:3] hack" will break for 2D data. Also, we might want to make sure the output arr_pred_ref_space
    #  always has the same shape as the input image. For multiple class prediction, maybe we can find another way
    #  to store the predictions (e.g. list of np arrays?).
~~~

jcohenadad replied:
> From this error https://github.com/ivadomed/ivadomed/pull/714/checks?check_run_id=2107966525, the problem seems to come from the save_color_labels function which generates nifti files with non-int or float data. Rather than introducing this condition that might cause errors with 2D images, I'd simply comment the lines related to this function (which simply creates an extra file that we barely use) and adjust the function save_color_labels in a future PR. If this plans work for you, I can take care of it.

that would be awesome! yes-- i am completely swamped with other tasks now

jcohenadad replied:
so IIUC: what you are suggesting is that after the fix you suggest, `arr_pred_ref_space.shape` will ALWAYS be the same as `nib_ref.get_fdata().shape`?  glancing at the code, my understanding was that in some cases the 4th dimension is used for different channels (so you suggest to not do that anymore?)

andreanne-lemay replied:
Yes, exactly! The prediction and the original image will always have different dimensions when using ivadomed as we store the different classes in the 4th dimension. Changing this would require an important refactoring. However, in SCT we always store the data in different files (one file per class) so this shape discrepancy should never happen when calling segment_volume from SCT. I'll do some test to verify if this is true. If so, I think only this line is required `nib_pred = nib.Nifti1Image(arr_pred_ref_space, None, nib_ref.header.copy())` since this code still works when the shapes are different (even though it might generate the qform and form codes, but with the current form of the code I think it is inevitable).

jcohenadad replied:
> since this code still works when the shapes are different

when i tested it, the following code:
~~~
nib_pred = nib.Nifti1Image(arr_pred_ref_space, None, nib_ref.header.copy())
~~~

will produce an error if e.g. one is (nx, ny, nz) and the other is (nx, ny, nz, 1). Hence my "hack"

ahill187 replied:
If there is a concern that this "[0:3] hack" will break for 2D data, should we add a test for 2D data?

andreanne-lemay replied:
I was wrong, the predictions generated from SCT initially have a shape of (nx, ny, nz, X) and are converted to a list afterward: https://github.com/ivadomed/ivadomed/blob/master/ivadomed/inference.py#L300. There will always be a shape discrepancy in the `pred_to_nib` function. 

I tested the code in its actual form (with the header copied even if the shapes are different), and I got the following qform/sform codes: 
![image](https://user-images.githubusercontent.com/49137243/111193368-79ca0680-8590-11eb-9913-c276bb5ca0f8.png)

The expected behavior is that qform and sform should always match the original image?

jcohenadad replied:
> The expected behavior is that qform and sform should always match the original image?

that is correct.

andreanne-lemay replied:
The current version of the code doesn't fix the issue regarding the qform and sform codes I believe. See suggestion below.

jcohenadad commented on ivadomed/inference.py:
```diff
@@ -314,7 +330,7 @@ def split_classes(nib_prediction):
     pred_list = []
     for c in range(pred.shape[-1]):
-        class_pred = nib.Nifti1Image(pred[..., c].astype('float32'), nib_prediction.affine)
+        class_pred = nib.Nifti1Image(pred[..., c].astype('float32'), None, nib_prediction.header)
```
need to copy to another object (to avoid pointer conflicts)
```suggestion
        class_pred = nib.Nifti1Image(pred[..., c].astype('float32'), None, nib_prediction.header.copy())
```

jcohenadad committed:
need to copy to another object (to avoid pointer conflicts)

ahill187 commented:
What's the reasoning for having the tests in a separate pull request?

jcohenadad commented:
> What's the reasoning for having the tests in a separate pull request?

because i don't have the time to do it 😊   but if pls feel free to chip in1

ahill187 commented:
> > What's the reasoning for having the tests in a separate pull request?
> 
> because i don't have the time to do it 😊 but if pls feel free to chip in1

Haha okay, I'll look into adding some tests then! :)

ahill187 committed:
Minor code cleaning in inference.py

ahill187 committed:
Rename test_inference to test_testing

Python testing standards typically organize tests by the files they
are testing. Since test_inference is testing the testing.py file,
it makes sense to name it accordingly.

andreanne-lemay commented:
@jcohenadad the easiest patch I can think of is doing something like this in the `segment_volume` function ([here](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/inference.py#L300)), instead of changing the code in `pred_to_nib` where the prediction will never have the same dimensions as the original image.

```               
                pred_list = split_classes(pred_nib)

                def change_header(pred_list, fname_ref):
                    """Change header to keep qform and sform codes
                    Args:
                        pred_list (list): List of predictions (with shape H,W,D)
                        fname_ref (str): Path to the reference image used to extract the header.
                    
                    Return:
                        list: List of prediction with updated header.                      
                    
                    """
                    nib_ref = nib.load(fname_ref)
                    pred_updated_list = []
                    print("CHANGE HEADER")
                    for pred in pred_list:
                        nib_pred = nib.Nifti1Image(pred.get_fdata(), None, nib_ref.header.copy())
                        pred_updated_list.append(nib_pred)
                    return pred_updated_list

                pred_list = change_header(pred_list, fname_images[0])
                target_list = context['loader_parameters']['target_suffix']
```

If you agree, with this change I can update the code in the branch. I tested it locally and it fixes the issue of the qform and sform codes.

andreanne-lemay commented:
> @jcohenadad the easiest patch I can think of is doing something like this in the `segment_volume` function ([here](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/inference.py#L300)), instead of changing the code in `pred_to_nib` where the prediction will never have the same dimensions as the original image.
> 
> ```
>                 pred_list = split_classes(pred_nib)
> 
>                 def change_header(pred_list, fname_ref):
>                     """Change header to keep qform and sform codes
>                     Args:
>                         pred_list (list): List of predictions (with shape H,W,D)
>                         fname_ref (str): Path to the reference image used to extract the header.
>                     
>                     Return:
>                         list: List of prediction with updated header.                      
>                     
>                     """
>                     nib_ref = nib.load(fname_ref)
>                     pred_updated_list = []
>                     print("CHANGE HEADER")
>                     for pred in pred_list:
>                         nib_pred = nib.Nifti1Image(pred.get_fdata(), None, nib_ref.header.copy())
>                         pred_updated_list.append(nib_pred)
>                     return pred_updated_list
> 
>                 pred_list = change_header(pred_list, fname_images[0])
>                 target_list = context['loader_parameters']['target_suffix']
> ```
> 
> If you agree, with this change I can update the code in the branch. I tested it locally and it fixes the issue of the qform and sform codes.

Ah 3290f85 this does the same as I what I suggested. I think only the code in `split_classes` should be change. I don't think we need to change the code in `pred_to_nib`.

jcohenadad commented:
> I think only the code in `split_classes` should be changed. 

`split_classes` takes as input `nib_prediction`:
https://github.com/ivadomed/ivadomed/blob/0a4b7bb3fad8092b31b347332b16f5308be09a54/ivadomed/inference.py#L306

Which means, that if the header of `nib_prediction` is wrong, then the wrong header will be propagated inside `split_classes` as well, even if we introduce the change in https://github.com/ivadomed/ivadomed/commit/3290f85b0f417af451fbfd3d8f07152ec3ea44f2.

I think we need to change only one line in `pred_to_nib`:

If you are suggesting to only change this line:
https://github.com/ivadomed/ivadomed/blob/0a4b7bb3fad8092b31b347332b16f5308be09a54/ivadomed/inference.py#L89
into this:
> `nib_pred = nib.Nifti1Image( dataobj=arr_pred_ref_space, affine=None, header=nib_ref.header.copy() )`

we will end up with the situation i am describing in https://github.com/ivadomed/ivadomed/pull/714/files#r594479677 (hence the hack).

I feel we are going in circles-- maybe a quick chat would be more efficient

jcohenadad committed:
Removed check on prediction shape

Prediction should always have the same nx,ny,nz as the reference image. A hack was introduced temporarily to fix a test, however it was decided to silence the test temporarily.
More details in: https://github.com/ivadomed/ivadomed/pull/714#pullrequestreview-612327091

jcohenadad committed:
Silenced a test

An issue will be opened to follow up

jcohenadad committed:
Silenced calls of save_color_labels.

More details in: https://github.com/ivadomed/ivadomed/issues/720

andreanne-lemay reviewed (approved):
LGTM!

gisellemartel commented on ivadomed/testing.py:
```diff
@@ -217,11 +220,13 @@ def run_inference(test_loader, model, model_params, testing_params, ofolder, cud
+                                       'https://github.com/ivadomed/ivadomed/issues/720')
+                        # imed_visualize.save_color_labels(np.stack(pred_tmp_lst, -1),
+                        #                              False,
+                        #                              fname_tmp,
```
Is there a reason we are leaving this commented out code here?

jcohenadad replied:
justified in eca1c5f089e1a3f5a5fefe248a7746af10ef7fff

gisellemartel commented on ivadomed/testing.py:
```diff
@@ -270,11 +275,13 @@ def run_inference(test_loader, model, model_params, testing_params, ofolder, cud
-                                                     slice_axis)
+                        logger.warning('No color labels saved due to a temporary issue. For more details see:'
+                                       'https://github.com/ivadomed/ivadomed/issues/720')
+                        # imed_visualize.save_color_labels(pred_undo,
+                        #                              False,
+                        #                              batch['input_metadata'][smp_idx][0]['input_filenames'],
+                        #                              fname_pred.split(".nii.gz")[0] + '_color.nii.gz',
+                        #                              slice_axis)
```
same as comment above

jcohenadad replied:
justified in justified in eca1c5f089e1a3f5a5fefe248a7746af10ef7fff

gisellemartel commented on testing/unit_tests/test_rgb.py:
```diff
@@ -9,23 +9,27 @@ def setup_function():
-        output_filename=os.path.join(__tmp_dir__, "rgb_test.nii.gz"),
-        slice_axis=0
-    )
+# TODO: un-silence this test. This test was silenced because it was failing after introducing the copy of the header in
```
If there is an issue ticket for this, it would be good practice to reference the issue # here

jcohenadad replied:
thanks, done in https://github.com/ivadomed/ivadomed/pull/714/commits/a00a54fc5c701ba6f717168c2ed43a972230b051

gisellemartel reviewed (approved):
Look's good! Just some minor suggestions for code quality

jcohenadad committed:
Merge branch 'master' into jca/711-qform

jcohenadad committed:
Added link to issue

jcohenadad committed:
Added link to issue

jcohenadad committed:
Merge branch 'master' into jca/711-qform
